### ISSUE TITLE
Difference in accuracy between tensorflow-GPU and tensorflow-CPU with 10x kfold

### ISSUE THREAD

**Issue by danielduckworth** [uber/ludwig#663]
**Describe the bug**
I've been running some trials comparing results of ludwig using tensorflow with GPU and CPU.

With an identical dataset and using 10x kfolds there are significant differences in the mean kfold accuracy.

On three separate datasets I got the following mean kfold accuracy with tensorflow-gpu:

0.903
0.893
0.801

On the same datasets I got the following results with tensorflow-cpu:

0.965
0.988
0.949

**To Reproduce**

For the GPU runs I install ludwig from github and modify the requirements.txt to use "tensorflow-gpu==1.15.2"

For the CPU runs I install ludwig using !pip install ludwig

**Expected behavior**
I expected the results to differ by no more than 1% - 2%.

**Environment (please complete the following information):**
 - OS: Google Colab
- Python version: 3.6
- Ludwig version: 0.2.2

**Additional context**

One thing I noticed was a slight difference in the version of tensorflow between GPU and CPU.

I ran the following:

import tensorflow as tf
print(tf.__version__)

For the GPU install of tensorflow the version is 1.15.0
For the CPU install of tensorflow the version is 1.15.2

**Comment by w4nderlust** [comment]
Thanks for this @danielduckworth . It's indeed pretty weird, I would have expected slightly different results (do to appeoximation / randomization) but these diefferences are substantial.
I will definitely look into it. Would it be possible for you to share an entirely replicable setting, with data, script / command and yaml file?
If not, would it be possible to try to replicate using the dataset synthesizer in `data/dataset_synthesizer.py`?
Also it could be a feature type dependent issue, so even if you can't provide data and are unable to replicate, at least knowing your yaml configuration would still be useful.

Finally, just for confirming the issue, could you please try using `tensorflow-gpu==1.15.2`?

**Comment by danielduckworth** [comment]
I can't share the data but the model definition is below.

I'll have a go at using the dataset synthesizer and see if I can reproduce the results.

**CLI commands:**

!ludwig experiment --data_csv "file.csv" --model_definition_file "model.yaml" --k_fold 10

**Model definition:**

training:
 dropout_rate: 0.2
 epochs: 50

input_features:
-
 name: response
 type: text
 encoder: rnn
 pretrained_embeddings: glove.6B.300d.txt
 embedding_size: 300
 cell_type: lstm_cudnn
 bidirectional: true
 dropout: true

output_features:
-
 name: score
 type: category

**Comment by danielduckworth** [comment]
@w4nderlust Do you have any examples for using the dataset synthesizer?

The dataset I want to emulate is about 2,000 text samples, with an average of 95 words and 3 categorical classes (0,1,2).

**Comment by w4nderlust** [comment]
By looking at your model, I have a suspition it may have to do with `lstm_cudnn`. Could you please try with regular `lstm` for both GPU and CPU? Also, again, making sure you have 1.15.2 installed for both.

> @w4nderlust Do you have any examples for using the dataset synthesizer?
> 
> The dataset I want to emulate is about 2,000 text samples, with an average of 95 words and 3 categorical classes (0,1,2).

Unfortunately I don't, will want to create one eventually. Anyway, if you run `-h` you'll see help explaining a bit. The most important feature is `--features` and here is the definition, the default is a good example i guess:
```
parser.add_argument(
        '-f',
        '--features',
        default='[\
          {name: text_1, type: text, vocab_size: 20, max_len: 20}, \
          {name: text_2, type: text, vocab_size: 20, max_len: 20}, \
          {name: category_1, type: category, vocab_size: 10}, \
          {name: category_2, type: category, vocab_size: 15}, \
          {name: numerical_1, type: numerical}, \
          {name: numerical_2, type: numerical}, \
          {name: binary_1, type: binary}, \
          {name: binary_2, type: binary}, \
          {name: set_1, type: set, vocab_size: 20, max_len: 20}, \
          {name: set_2, type: set, vocab_size: 20, max_len: 20}, \
          {name: bag_1, type: bag, vocab_size: 20, max_len: 10}, \
          {name: bag_2, type: bag, vocab_size: 20, max_len: 10}, \
          {name: sequence_1, type: sequence, vocab_size: 20, max_len: 20}, \
          {name: sequence_2, type: sequence, vocab_size: 20, max_len: 20}, \
          {name: timeseries_1, type: timeseries, max_len: 20}, \
          {name: timeseries_2, type: timeseries, max_len: 20}, \
          ]',
        type=yaml.safe_load, help='dataset features'
    )
```

**Comment by danielduckworth** [comment]
Hi @w4nderlust looks like you were right.

Running the model with just 'lstm' as the cell type fixed the issue.

Same datasets with GPU (without cudnn)

0.962
0.988
0.949

Almost exactly the same as the original CPU runs.

It has also solved what I thought was a second problem which was that for some dataset (not the three above) some folders were getting an accuracy of 0. I thought this might be a class imbalance issues but it is no longer happening.

In Google Colab the default install of Tensorflow is 1.5.0 which I left based on Google's recommendation in their documentation rather than 1.5.2 which you suggested.

I also haven't run the trial again with the CPU install but I can if you think it

Thanks for your help with this issue.

Also, the data sythensizer is excellent! It's going to be really useful for testing.

**Comment by w4nderlust** [comment]
@danielduckworth glad to hear this!
It would be grat if you could run a apple to apple test with tf 1.15.2, but don't worry if that's not feasible.
So closing this issue, please reopen if other issues pop up.

And yes, the dataset synthesizer is pretty cool :) I was planning on writing some documentation for it but never found the time so far. If by any chance you have the time to write a mini guide on how to use it from you own experience, that would be greatly appreciated! :)
